### PR TITLE
Add read only transactions

### PR DESCRIPTION
**Pull request by jjestrel**
https://github.com/square/misk/issues/223

### CODE REVIEW

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 1]
```diff
@@ -162,12 +170,24 @@ internal class RealTransacter private constructor(
...
-    val config: DataSourceConfig
-  ) : Session {
+  internal class ReadOnlySession(
+    session: org.hibernate.Session,
+    config: DataSourceConfig
+  ) : BaseSession(session, config) {
     override val hibernateSession = session
+    init {
+      hibernateSession.isDefaultReadOnly = true
+      hibernateSession.hibernateFlushMode = FlushMode.MANUAL
```
Comment (jjestrel) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt]: Hibernate says to do this for performance benefit when being read only

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 2]
```diff
@@ -162,12 +170,24 @@ internal class RealTransacter private constructor(
...
+    }
+    override fun <T : DbEntity<T>> save(entity: T): Id<T> {
+      throw IllegalStateException("Saving isn't permitted in a read only session.")
+    }
+  }
 
+  internal class WritableSession(
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt]: lets chat about this . . .  I think readonly/readwrite is a great idea, but not a good fit for our type hierarchy

File: misk-hibernate/src/test/kotlin/misk/hibernate/TransacterTest.kt [hunk 3]
```diff
@@ -186,6 +186,35 @@ class TransacterTest {
     assertThat(callCount.get()).isEqualTo(1)
   }
 
+  @Test
+  fun readOnlyWontSave() {
```
Comment (swankjesse) [inline misk-hibernate/src/test/kotlin/misk/hibernate/TransacterTest.kt]: nice

File: misk-hibernate/src/test/kotlin/misk/hibernate/TransacterTest.kt [hunk 4]
```diff
@@ -186,6 +186,35 @@ class TransacterTest {
...
+        session.save(DbMovie("Star Wars", LocalDate.of(1977, 5, 25)))
+      }
+    }
+    transacter.transaction {session ->
+      assertThat(queryFactory.newQuery<MovieQuery>().list(session)).isEmpty()
+    }
+  }
+
+  @Test
+  fun readOnlyWontModify() {
```
Comment (swankjesse) [inline misk-hibernate/src/test/kotlin/misk/hibernate/TransacterTest.kt]: nice

**Comment by ryanhall07** [comment]
we should include documentation that this is different from using a datasource that is in readOnly mode. that seems to always cause confusion with developers.

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt (hunk 2 above)
Comment (jjestrel) [reply misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt]: Switched. We talked about layers of code that need to be modified if something is readonly/readwite, and List.remove.

**Comment by jjestrel** [comment]
Added doc for `readOnly()` with `@throws` inside too

File: misk-hibernate/src/main/kotlin/misk/hibernate/Transacter.kt [hunk 5]
```diff
@@ -23,6 +23,14 @@ interface Transacter {
   fun retries(): Transacter
 
   fun noRetries(): Transacter
+
+  /**
+   * Marks the sessions produced by this transacter as read only. This does not mean the underlying
+   * datasource is read only. Only that the session produced won't modify the database.
```
[the remark this answers is not in the document]
Comment (jjestrel) [reply misk-hibernate/src/main/kotlin/misk/hibernate/Transacter.kt]: We create a new Transacter like with the other options, will update

File: misk-hibernate/src/main/kotlin/misk/hibernate/Transacter.kt [hunk 6]
```diff
@@ -23,6 +23,14 @@ interface Transacter {
...
+   *
+   * @throws IllegalStateException when save is called on sessions produced by this Transacter
```
[the remark this answers is not in the document]
Comment (jjestrel) [reply misk-hibernate/src/main/kotlin/misk/hibernate/Transacter.kt]: Moved to Session.kt's `save`

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 7]
```diff
@@ -164,12 +168,22 @@ internal class RealTransacter private constructor(
...
+    init {
+      if (readOnly) {
+        hibernateSession.isDefaultReadOnly = true
+        hibernateSession.hibernateFlushMode = FlushMode.MANUAL
+      }
+    }
 
     @Suppress("UNCHECKED_CAST")
     override fun <T : DbEntity<T>> save(entity: T): Id<T> {
+      if (readOnly) {
```
Comment (ryanhall07) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt]: just wanted to double check that this check is only for redundancy and providing a nice error message. I'm hoping the above hibernate settings are properly flipping the readOnly bit on the jdbc connection. Is that true?

**Pull request merged**